src/recommendation/src/main/scala/RankingTrainValidationSplit.scala

```diff
@@ -0,0 +1,343 @@
+
+  /** @group setParam */
+  def setSeed(value: Long): this.type = set(seed, value)
+
+  /** @group setParam */
+  def setMinRatingsU(value: Int): this.type = set(minRatingsU, value)
+
+  /** @group setParam */
+  def setMinRatingsI(value: Int): this.type = set(minRatingsI, value)
+
+  override def transformSchema(schema: StructType): StructType = transformSchemaImpl(schema)
+
+  /**
+    * The number of threads to use when running parallel algorithms.
+    * Default is 1 for serial execution
+    *
+    * @group expertParam
+    */
+  val parallelism = new IntParam(this, "parallelism",
+    "the number of threads to use when running parallel algorithms", ParamValidators.gtEq(1))
+
+  setDefault(parallelism -> 1)
+
+  /** @group expertGetParam */
+  def getParallelism: Int = $(parallelism)
+
+  /** @group expertSetParam */
+  def setParallelism(value: Int): this.type = set(parallelism, value)
+
+  private[ml] def getExecutionContext: ExecutionContext = {
+
+    getParallelism match {
+      case 1 =>
+        SparkHelpers.getThreadUtils().sameThread
+      case n =>
+        ExecutionContext.fromExecutorService(SparkHelpers.getThreadUtils()
+          .newDaemonCachedThreadPool(s"${this.getClass.getSimpleName}-thread-pool", n))
+    }
+  }
+
+  override def fit(dataset: Dataset[_]): RankingTrainValidationSplitModel = {
+    val schema = dataset.schema
+    transformSchema(schema, logging = true)
+    val est = $(estimator)
+    val eval = $(evaluator).asInstanceOf[RankingEvaluator]
+    val epm = $(estimatorParamMaps)
+    val numModels = epm.length
+
+    dataset.cache()
+    eval.setNItems(dataset.agg(countDistinct(col($(itemCol)))).take(1)(0).getLong(0))
+    val filteredDataset = filterRatings(dataset.dropDuplicates())
+
+    //Stratified Split of Dataset
+    val Array(trainingDataset, validationDataset): Array[DataFrame] = splitDF(filteredDataset)
+    trainingDataset.cache()
+    validationDataset.cache()
+
+    val executionContext = getExecutionContext
+
+    def calculateMetrics(model: Transformer, validationDataset: Dataset[_]): Double = model match {
+      case p: PipelineModel =>
+        //Assume Rec Algo is last stage of pipeline
+        val modelTemp = model.asInstanceOf[PipelineModel].stages.last
+        calculateMetrics(modelTemp, validationDataset)
+      case a: ALSModel =>
+        val recs = model.asInstanceOf[ALSModel].recommendForAllUsers(5 * eval.getK)
+        val preparedTest: Dataset[_] = prepareTestData(validationDataset.toDF(), recs, eval.getK)
+        eval.evaluate(preparedTest)
+    }
+
+    val metricFutures = epm.zipWithIndex.map { case (paramMap, paramIndex) =>
+      Future[Double] {
+        val model = est.fit(trainingDataset, paramMap).asInstanceOf[Model[_]]
+        calculateMetrics(model, validationDataset)
+      }(executionContext)
+    }
+
+    val metrics = metricFutures.map(SparkHelpers.getThreadUtils().awaitResult(_, Duration.Inf))
```

Comment:
does spark use an infinite timeout for this?

Reply:
![image](https://user-images.githubusercontent.com/9027725/50833823-77205680-1320-11e9-83a5-f325cedc9098.png)

Reply:
yup.